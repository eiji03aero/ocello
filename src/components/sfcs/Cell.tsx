import * as React from "react";
import styled from "styled-components";

const CellStyled = styled.div`
  background-color: green;
  display: inline-block;
  width: 80px;
  height: 80px;
  margin-bottom: 10px;
  margin-right: 10px;
`;

interface Props {
  children: React.ReactNode;
}

export const Cell: React.SFC = (props: Props) => {
  return (
    <CellStyled>
      { props.children }
    </CellStyled>
  );
};
