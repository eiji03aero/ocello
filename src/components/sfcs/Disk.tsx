import * as React from "react";
import * as cn from "classnames";
import styled from "styled-components";

import { colors } from "../../utils/colors";

const DiskStyled = styled.div`
  width: 85%;
  height: 85%;
  border-radius: 50%;
  background-color: transparent;

  &.Black {
    background-color: ${colors.black};
  }

  &.White {
    background-color: ${colors.white};
  }
`;

interface Props {
  classNames: any;
}

export const Disk: React.SFC<Props> = ({
  classNames,
}: Props) => {
  return (
    <DiskStyled
      className={cn(classNames)}
    />
  );
}
