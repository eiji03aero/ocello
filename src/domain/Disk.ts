export enum DiskColors {
  Black = "Black",
  White = "White",
}

export interface DiskArgs {
  color: DiskColors;
}

export class Disk {
  color: DiskColors;

  constructor (args: DiskArgs) {
    this.color = args.color;
  }

  static Black () {
    return new Disk({ color: DiskColors.Black });
  }

  static White () {
    return new Disk({ color: DiskColors.White });
  }
}
