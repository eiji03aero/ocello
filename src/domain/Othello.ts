import { Board } from "./Board";

export enum GameStatus {
  PLAYING,
  FINISHED,
}

export enum UserTurn {
  ONE,
  TWO,
}

export interface OthelloArgs {
  board: Board;
  gameStatus: GameStatus;
  userTurn: UserTurn;
}

export class Othello {
  board: Board;
  gameStatus: GameStatus;
  userTurn: UserTurn;

  constructor (args: OthelloArgs) {
    this.board = args.board;
    this.gameStatus = args.gameStatus;
    this.userTurn = args.userTurn;
  }

  static NewGame () {
    return new Othello({
      board: Board.Blank(),
      gameStatus: GameStatus.PLAYING,
      userTurn: UserTurn.ONE,
    });
  }

  static RandomGame () {
    return new Othello({
      board: Board.Random(),
      gameStatus: GameStatus.PLAYING,
      userTurn: UserTurn.ONE,
    });
  }
}
