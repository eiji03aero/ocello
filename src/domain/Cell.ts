import { Disk, DiskColors } from "./Disk";

export enum CellState {
  Blank = "Blank",
  Placed = "Placed",
}

interface CellArgs {
  state: CellState;
  disk?: Disk;
}

export class Cell {
  state: CellState;
  disk: Disk | null;

  constructor (args: CellArgs) {
    this.state = args.state || CellState.Blank;
    this.disk = args.disk || null;
  }

  static Blank () {
    return new Cell({ state: CellState.Blank });
  }

  static Black () {
    return new Cell({ state: CellState.Placed, disk: Disk.Black() });
  }

  static White () {
    return new Cell({ state: CellState.Placed, disk: Disk.White() });
  }

  static fromColor (color: DiskColors) {
    return new Cell({ state: CellState.Placed, disk: new Disk({ color: color })});
  }

  get color () {
    return this.state === CellState.Placed
      ? this.disk.color
      : '';
  }

  get isBlank () {
    return this.state === CellState.Blank;
  }

  get isPlaced () {
    return this.state === CellState.Placed;
  }
}
