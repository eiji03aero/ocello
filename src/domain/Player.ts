import { DiskColors } from "./Disk";

export interface PlayerArgs {
  color: DiskColors;
}

export class Player {
  color: DiskColors;

  constructor (args: PlayerArgs) {
    this.color = args.color;
  }

  static Black () { return new Player({ color: DiskColors.Black }); }
  static White () { return new Player({ color: DiskColors.White }); }
}
