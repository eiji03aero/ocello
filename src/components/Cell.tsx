import * as React from "react";
import * as cn from "classnames";
import styled from "styled-components";

import { Disk } from "./Disk";

import { CellCoordinates } from "../domain/Board";
import { Cell as CellClass } from "../domain/Cell";
import { colors } from "../utils/colors";

const CellStyled = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 80px;
  height: 80px;
  background-color: ${colors.green};
  border: 1px solid ${colors.black};

  &.placeable {
    &:hover {
      cursor: pointer;
      opacity: .8;
    }
  }

  &.not-placeable {
    &:hover {
      pointer-events: none;
    }
  }
`;

export interface Props {
  cell: CellClass;
  rowIndex: number;
  columnIndex: number;
  onPlaceDisk: (coords: CellCoordinates) => void;

  placeable: boolean;
}

export const Cell: React.SFC<Props> = ({
  cell,
  rowIndex,
  columnIndex,
  onPlaceDisk,

  placeable,
}: Props) => {
  const classNames = cn(placeable ? 'placeable' : 'not-placeable')
  const onClickHandler = (e: React.MouseEvent) => {
    if (placeable) {
      onPlaceDisk([rowIndex, columnIndex]);
    }
  }

  return (
    <CellStyled className={classNames}
      onClick={onClickHandler}
    >
      <Disk classNames={cell.color} />
    </CellStyled>
  );
};
