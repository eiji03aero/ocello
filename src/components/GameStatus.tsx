import * as React from "react";
import styled from "styled-components";

import { Player } from "../domain/Player";
import { BoardStatus } from "../domain/Board";
import { colors } from "../utils/colors";

const GameStatusStyled = styled.div`
  display: inline-block;
  padding: 8px 0 4px;

  .box {
    padding-left: 4px;
    border-left: 4px solid ${colors.green};
  }

  .box + .box {
    margin-left: 16px;
  }
`;

export interface Props {
  currentPlayer: Player;
  boardStatus: BoardStatus;
  canSkip: boolean;
  onSkipTurn: (e: React.SyntheticEvent) => void;
}

export const GameStatus: React.SFC<Props> = ({
  currentPlayer,
  boardStatus,
  canSkip,
  onSkipTurn,
}: Props) => {
  return (
    <GameStatusStyled>
      <span className="box">
        Current player: { currentPlayer.color }
      </span>
      <span className="box">
        Black: {boardStatus.Black}
      </span>
      <span className="box">
        White: {boardStatus.White}
      </span>
      <span className="box">
        Blank: {boardStatus.Blank}
      </span>
      { canSkip && (
        <button onClick={onSkipTurn}>
          Skip turn
        </button>
      )}
    </GameStatusStyled>
  );
};
