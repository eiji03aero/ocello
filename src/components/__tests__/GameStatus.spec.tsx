import * as React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import * as _ from "lodash";

import { GameStatus } from "../GameStatus";

import { Player } from "../../domain/Player";
import { CellState } from "../../domain/Cell";
import { DiskColors } from "../../domain/Disk";
import { GameStatus as GameStatusEnum } from "../../domain/Othello";

describe("GameStatus", () => {
  const defaultProps = {
    currentPlayer: Player.Black(),
    boardStatus: {
      [DiskColors.White]: 2,
      [DiskColors.Black]: 2,
      [CellState.Blank]: 60,
    },
    gameStatus: GameStatusEnum.PLAYING,
    onSkipTurn: (e: React.SyntheticEvent) => {},
    onNewGame: (e: React.SyntheticEvent) => {},
  };

  it("matches snapshot", () => {
    const wrapper = shallow( <GameStatus { ...defaultProps } />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
