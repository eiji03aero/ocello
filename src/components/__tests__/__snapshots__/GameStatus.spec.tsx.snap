// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameStatus matches snapshot 1`] = `
<styled.div>
  <span
    className="box"
  >
    Current player: 
    Black
  </span>
  <span
    className="box"
  >
    Black: 
    2
  </span>
  <span
    className="box"
  >
    White: 
    2
  </span>
  <span
    className="box"
  >
    Blank: 
    60
  </span>
  <button
    onClick={[Function]}
  >
    New game
  </button>
  <button
    disabled={true}
    onClick={[Function]}
  >
    Skip turn
  </button>
</styled.div>
`;
